### PR TITLE
Refactor / improve replaceAllMemRefUsesWith

### PR DESCRIPTION
**Pull request by bondhugula**
Refactor replaceAllMemRefUsesWith to split it into two methods: the new
method does the replacement on a single op, and is used by the existing
one.

- make the methods return LogicalResult instead of bool

- Earlier, when replacement failed (due to non-deferencing uses of the
  memref), the set of ops that had already been processed would have
  been replaced leaving the IR in an inconsistent state. Now, a
  pass is made over all ops to first check for non-deferencing
  uses, and then replacement is performed. No test cases were affected
  because all clients of this method were first checking for
  non-deferencing uses before calling this method (for other reasons).
  This isn't true for a use case in another upcoming PR (scalar
  replacement); clients can now bail out with consistent IR on failure
  of replaceAllMemRefUsesWith.

- multiple deferencing uses of the same memref in a single op is
  possible (we have no such use cases/scenarios), and this has always
  remained unsupported. Add an assertion for this.

Signed-off-by: Uday Bondhugula <uday@polymagelabs.com>

### CODE REVIEW

File: include/mlir/Transforms/Utils.h [hunk 1]
```diff
@@ -66,12 +66,20 @@ class OpBuilder;
...
-                              Operation *postDomInstFilter = nullptr);
+LogicalResult replaceAllMemRefUsesWith(Value *oldMemRef, Value *newMemRef,
+                                       ArrayRef<Value *> extraIndices = {},
+                                       AffineMap indexRemap = AffineMap(),
+                                       ArrayRef<Value *> extraOperands = {},
+                                       Operation *domInstFilter = nullptr,
+                                       Operation *postDomInstFilter = nullptr);
+
+/// Performs the same replacement as the other version above but only for the
+/// deferencing uses of `oldMemRef` in `op`.
```
Comment (joker-eph) [inline include/mlir/Transforms/Utils.h]: I am not sure what  "deferencing uses of `oldMemRef` in `op`." means?

File: lib/Transforms/Utils/Utils.cpp [hunk 2]
```diff
@@ -57,12 +57,178 @@ static NamedAttribute getAffineMapAttrForMemRef(Operation *op, Value *memref) {
...
+    assert(indexRemap.getNumResults() + extraIndices.size() == newMemRefRank);
+  } else {
+    assert(oldMemRefRank + extraIndices.size() == newMemRefRank);
+  }
+
+  // Assert same elemental type.
+  assert(oldMemRef->getType().cast<MemRefType>().getElementType() ==
+         newMemRef->getType().cast<MemRefType>().getElementType());
+
+  // Check if the memref was used in a non-deferencing context.
```
Comment (joker-eph) [inline lib/Transforms/Utils/Utils.cpp]: `deferencing` -> `dereferencing` (in many places in this PR)

Also the comment is literally paraphrasing what the next statement is, which is quite an anti-pattern.

File: include/mlir/Transforms/Utils.h (hunk 1 above)
Comment (bondhugula) [reply include/mlir/Transforms/Utils.h]: A dereferencing use is one where the memref is dereferenced, i.e., an individual element of it is accessed. For eg:

%v =  load %A[%i, %j] : memref<100 x 100 x f32> 

is a deferencing use of %A. affine.load, affine.store, load, store, dma_start, dma_finish are examples of ops that have deferencing uses of the memref. OTOH, uses in ops like:

%v = call "foo" (%A) : memref<100 x 100 x f32>
dealloc %A : memref<100 x 100 x f32>

are all non-deferencing uses of the memref. The doc comment is merely meant as a indirect reference to the version of the method above, which has all the info.

Comment (bondhugula) [reply include/mlir/Transforms/Utils.h]: Sorry, just realized there was a typo! Thanks.

**Comment by bondhugula** [comment]
> We should be able to expose such fix independently through tests: can one of the test passes be improved to expose this?

None of the current users of this method are actually affected by this fix, but there is another PR that I am about to submit on scalar replacement which relies on this (i.e., replaceAllMemRefUsesWith can fail and the failure is handled), and have a test case for that. However that PR is too big in itself and separate from this. 

The reason the existing users always make sure that this methods succeeds is that replaceAllMemRefUsesWith is typically called by a pass/utility after a bunch of IR actions have already been performed (new op creation, etc.). So, if it were to fail, one would have to undo/clear up all of those actions. As a result, those clients always first check from the outside a-priori whether it'd succeed (by checking for non-dereferencing uses) and then call it.

File: lib/Transforms/Utils/Utils.cpp (hunk 2 above)
Comment (bondhugula) [reply lib/Transforms/Utils/Utils.cpp]: Fixed.

File: include/mlir/Transforms/Utils.h (hunk 1 above)
Comment (bondhugula) [reply include/mlir/Transforms/Utils.h]: Fixed.

**Comment by bondhugula** [comment]
Just found a way to test this - PipelineDataTransfer does check for and handle the failure. Added a test case - with this change, the IR remains correct after the replacement failure. 

Thanks!

**Comment by antiagainst** [comment]
Hey @joker-eph or @River707, could you take another look over this PR?

File: test/Transforms/pipeline-data-transfer.mlir [hunk 3]
```diff
@@ -52,7 +52,7 @@ func @loop_nest_dma() {
 // CHECK-NEXT:  %{{.*}} = affine.load %{{.*}}[%{{.*}} mod 2, %{{.*}}] : memref<2x32xf32, 1>
 // CHECK-NEXT:  %{{.*}} = "compute"(%{{.*}}) : (f32) -> f32
 // CHECK-NEXT:  affine.store %{{.*}}, %{{.*}}[%{{.*}} mod 2, %{{.*}}] : memref<2x32xf32, 1>
-// CHECK-NEXT:  affine.for %{{.*}} = 0 to 128 {
+// CHECK-NEXT:  affine.for %{{.*}} = 0 to 32 {
```
Comment (joker-eph) [inline test/Transforms/pipeline-data-transfer.mlir]: Why these changes 128->32?

Comment (bondhugula) [reply test/Transforms/pipeline-data-transfer.mlir]: This one is minor and orthogonal. The DMA buffer size / transfer size in the op's above is actually 32, and so the loop should be of trip count 32 to be meaningful. It doesn't affect this test case - but just fixed it while on this file.

File: lib/Transforms/PipelineDataTransfer.cpp [hunk 4]
```diff
@@ -117,11 +117,12 @@ static bool doubleBuffer(Value *oldMemRef, AffineForOp forOp) {
 
   // replaceAllMemRefUsesWith will always succeed unless the forOp body has
   // non-deferencing uses of the memref (dealloc's are fine though).
```
[the remark this answers is not in the document]
Comment (bondhugula) [reply lib/Transforms/PipelineDataTransfer.cpp]: Fixed, thanks.

**Comment by bondhugula** [comment]
Anything more on this?

File: lib/Transforms/Utils/Utils.cpp [hunk 5]
```diff
@@ -89,170 +254,44 @@ bool mlir::replaceAllMemRefUsesWith(Value *oldMemRef, Value *newMemRef,
...
+    if (postDomInstFilter && !postDomInfo->postDominates(postDomInstFilter, op))
       continue;
 
-    // Skip dealloc's - no replacement is necessary, and a replacement doesn't
-    // hurt dealloc's.
-    if (isa<DeallocOp>(opInst))
+    // Skip dealloc's - no replacement is necessary, and a memref replacement
+    // at other uses doesn't hurt these dealloc's.
+    if (isa<DeallocOp>(op))
       continue;
```
[the remark this answers is not in the document]
Comment (bondhugula) [reply lib/Transforms/Utils/Utils.cpp]: Looks like this PR has already been committed. I'll address all of these coverage issues in a follow-up.

File: lib/Transforms/Utils/Utils.cpp [hunk 6]
```diff
@@ -57,12 +57,177 @@ static NamedAttribute getAffineMapAttrForMemRef(Operation *op, Value *memref) {
...
+  }
+
+  // Assert same elemental type.
+  assert(oldMemRef->getType().cast<MemRefType>().getElementType() ==
+         newMemRef->getType().cast<MemRefType>().getElementType());
+
+  if (!isMemRefDereferencingOp(*op))
+    // Failure: memref used in a non-dereferencing context (potentially
+    // escapes); no replacement in these cases.
+    return failure();
```
[the remark this answers is not in the document]
Comment (bondhugula) [reply lib/Transforms/Utils/Utils.cpp]: This one does have test coverage - from lib/Transforms/PipelineDataTransfer.cpp and tested via:
https://github.com/tensorflow/mlir/blob/c340184d48a9650d6606d193cce4fdff127e821e/test/Transforms/pipeline-data-transfer.mlir#L220

File: lib/Transforms/Utils/Utils.cpp [hunk 7]
```diff
@@ -57,12 +57,177 @@ static NamedAttribute getAffineMapAttrForMemRef(Operation *op, Value *memref) {
...
+
+  SmallVector<unsigned, 2> usePositions;
+  for (const auto &opEntry : llvm::enumerate(op->getOperands())) {
+    if (opEntry.value() == oldMemRef)
+      usePositions.push_back(opEntry.index());
+  }
+
+  // If memref doesn't appear, nothing to do.
+  if (usePositions.empty())
+    return success();
```
[the remark this answers is not in the document]
Comment (bondhugula) [reply lib/Transforms/Utils/Utils.cpp]: The current user of this method (only one) doesn't need/cover this case, but there is an upcoming scalar replacement utility for memref's that uses this method directly. Putting both in the same PR would have made it too big.

[the remark this answers is not in the document]
Comment (bondhugula) [reply lib/Transforms/Utils/Utils.cpp]: The current user of this method (only one) doesn't need/cover this case, but there is a scalar replacement utility for memref's that I plan to submit soon that uses this method directly. Putting both in the same PR would have made it too big.

File: lib/Transforms/Utils/Utils.cpp [hunk 8]
```diff
@@ -89,170 +254,44 @@ bool mlir::replaceAllMemRefUsesWith(Value *oldMemRef, Value *newMemRef,
...
+    // that has the use, and that op could be postDomFilter or domFilter itself!
+    opsToReplace.insert(op);
   }
 
-  for (auto *opInst : opsToErase)
-    opInst->erase();
+  for (auto *op : opsToReplace) {
+    if (failed(replaceAllMemRefUsesWith(oldMemRef, newMemRef, op, extraIndices,
+                                        indexRemap, extraOperands)))
+      assert(false && "memref replacement guaranteed to succeed here");
```
[the remark this answers is not in the document]
Comment (bondhugula) [reply lib/Transforms/Utils/Utils.cpp]: Yes, will change this in a follow-up PR.

File: test/Transforms/pipeline-data-transfer.mlir (hunk 3 above)
Comment (bondhugula) [reply test/Transforms/pipeline-data-transfer.mlir]: Sure. (I was under the impression that cleanup/NFC in the same file was fine.)

File: lib/Transforms/Utils/Utils.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (bondhugula) [reply lib/Transforms/Utils/Utils.cpp]: You are right. I confused this one with the same check in the other version of this method. This will be covered only when another PR that actually uses this method directly (scalar replacement) is submitted.
